src/Filters/stream_filt.jl

```diff
@@ -218,15 +218,19 @@ end
 
 function setphase!(kernel::Union{FIRInterpolator, FIRRational}, ϕ::Real)
     ϕ >= zero(ϕ) || throw(ArgumentError("ϕ must be >= 0"))
-    (ϕ, xThrowaway) = modf(ϕ)
-    kernel.inputDeficit += round(Int, xThrowaway)
-    kernel.ϕIdx = round(ϕ*(kernel.Nϕ) + 1.0)
+    xThrowaway, ϕIdx = divrem(round(Int, ϕ * kernel.Nϕ), kernel.Nϕ)
+    kernel.inputDeficit += xThrowaway
+    kernel.ϕIdx = ϕIdx + 1
     nothing
 end
 
 function setphase!(kernel::FIRArbitrary, ϕ::Real)
     ϕ >= zero(ϕ) || throw(ArgumentError("ϕ must be >= 0"))
     (ϕ, xThrowaway) = modf(ϕ)
+    if round(ϕ*kernel.Nϕ) == kernel.Nϕ
+        xThrowaway += 1
+        ϕ -= 1
```

Comment:
Good question. First: No, `ϕ` can become negative here. The `modf` ensures `0 <= ϕ < 1`, so after subtracting 1 it will become negative. In particular, we will have `-1/(2Nϕ) <= ϕ < 1-1/(2Nϕ)`.

What has to be maintained throughout is that `ϕIdx` is within `1:Nϕ`. Here it is initialized to `round(ϕ*Nϕ + 1.0)` which will fulfill the condition. (Without the subtraction, it could sometimes give `Nϕ+1` which will now wrap around to 1.) Further note that `ϕAccumulator` is initialized to `ϕ*Nϕ + 1.0`, and we will obviously then fulfill `0.5 <= ϕAccumulator < Nϕ+0.5`, where prior to this change, the bounds were 1 and `Nϕ+1`.

Now consider the `update!` function which is invoked for every output sample produced:
https://github.com/JuliaDSP/DSP.jl/blob/56773aac8783c491b637577624aaa62f788d97fe/src/Filters/stream_filt.jl#L570-L580
Now for the worst-case, if `ϕAccumulator == 0.5` and `kernel.Δ < 0.5`, the `floor` operation will give 0 and we run into an error. But can `kernel.Δ < 0.5` happen? It is set to `Nϕ/rate`, so yes, if `Nϕ < 0.5 * rate`. And one can specify both upon construction, so it is possible to hit this case.

So this change may indeed lead to a new error - but only in cases that would have errored also without this change, just even earlier. I'm unsure we should land this as is or reduce this PR to only the first commit and worry about `FIRArbitrary` separately.

Finally, I think there is something fishy going on as the initialization does
https://github.com/JuliaDSP/DSP.jl/blob/56773aac8783c491b637577624aaa62f788d97fe/src/Filters/stream_filt.jl#L232
and the update does
https://github.com/JuliaDSP/DSP.jl/blob/56773aac8783c491b637577624aaa62f788d97fe/src/Filters/stream_filt.jl#L578
and it is this discrepancy that prevents this fix to work in all cases. I think they should either both be `floor` or both be `round` but so far, I lack understanding of the inner workings of `FIRArbitrary` to decide. But if it's `floor`, the fix proposed here would need adjustment, so maybe that's a reason to focus on `Union{FIRInterpolator, FIRRational}` here.

Reply:
As an example for a situation not fixed by this PR:
```julia
julia> let H = FIRFilter(122.2)
           setphase!(H, 0.99)
           @test length(filt(H, 1:2)) == 3
       end
Error During Test at REPL[2]:3
  Test threw exception
  Expression: length(filt(H, 1:2)) == 3
  BoundsError: attempt to access 38×32 Matrix{Float64} at index [1, 33] # master
  BoundsError: attempt to access 38×32 Matrix{Float64} at index [1, 0] # this PR
```

Reply:
Ok, thinking about this some more, I really think that `setphase!` should be consistent with `update!` here and use `floor`. Consider this:
```
julia> let H = FIRFilter([0, 0, 1], 1000.1)
           setphase!(H, 0.03)
           filt(H, [1.0])[1:10]
       end
10-element Vector{Float64}:
 0.96
 0.0
 0.02399360063993594
 0.05599040095990393
 0.08798720127987192
 0.1199840015998399
 0.1519808019198079
 0.18397760223977588
 0.21597440255974387
 0.24797120287971186
```
(The following samples ramp further up.) The initial 0.96 looks quite off.  And indeed, replacing the `round` with `floor`, we get
```julia
julia> let H = FIRFilter([0, 0, 1], 1000.1)
           setphase!(H, 0.03)
           filt(H, [1.0])[1:10]
       end
10-element Vector{Float64}:
 0.0
 0.0
 0.02399360063993594
 0.05599040095990393
 0.08798720127987192
 0.1199840015998399
 0.1519808019198079
 0.18397760223977588
 0.21597440255974387
 0.24797120287971186
```
Ah, better. And considering that what `setphase!` stores in `kernel.ϕIdx` only affects the first sample, this is probably the way to go. I'll update the PR accordingly.